tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/RecordingZ3SolverContext.scala

```diff
@@ -0,0 +1,207 @@
+package at.forsyte.apalache.tla.bmcmt.smt
+
+import at.forsyte.apalache.tla.bmcmt.profiler.SmtListener
+import at.forsyte.apalache.tla.bmcmt.{Arena, ArenaCell}
+import at.forsyte.apalache.tla.lir.TlaEx
+import com.typesafe.scalalogging.LazyLogging
+
+
+object RecordingZ3SolverContext {
+  def apply(parentLog: Option[SmtLog], config: SolverConfig): RecordingZ3SolverContext = {
+    val context = new RecordingZ3SolverContext(parentLog, config)
+    parentLog match {
+      case Some(log) =>
+        log.replay(context.solver)
+
+      case None => ()
+    }
+    context
+  }
+}
+
+/**
+  * A wrapper around SolverContext that records SMT operations. The log is organized hierarchically,
+  * so a differential diff may be taken.
+  *
+  * @param parentLog the log of the parent context
+  * @param config solver configuration
+  *
+  * @author Igor Konnov
+  */
+@SerialVersionUID(700L)
+class RecordingZ3SolverContext private (parentLog: Option[SmtLog], val config: SolverConfig)
+    extends SolverContext with Serializable with LazyLogging {
+
+  import SmtLog._
+
+  private val solver = new Z3SolverContext(config)
+
+  /**
+    * The sequence of logs (the last added element is in the head),
+    * one per context, except the last log, which is maintained in lastLog.
```

Comment:
I don't see any `lastLog` in this module. Am I missing it? Or does doc need a tweak? Should this be `lastLogRev`?